### DISCUSSION TITLE
MacOS Middle Button

### DISCUSSION THREAD

**Discussion by Anzus** [Interlisp/medley#1099] [General]
OK, I’m coming back to the system after admittedly too long away, and the only way I can get middle button to work is by emulating 3-button in the XQuartz settings and using command-click. Didn’t middle button work at one point? Does it work in the SDL version?

**Comment by nbriggs** [comment]
We've been using emulate 3-button mouse in XQuartz.  There's also Lisp's emulated middle button, but it's holding down left and right at the same time, and I don't think you can do that in XQuartz, at least on a trackpad.  It's the same in SDL because it's the OS that decides whether you can do both buttons at the same time (if you plug in a real 2- or 3- button USB mouse it shouldn't be a problem)

**Comment by Anzus** [comment]
I should be clearer. I have a Logitech bluetooth trackball, which has 3 buttons (well, two buttons and a scroll wheel/button combo). Clicking that middle button doesn't do anything in Medley running in XQuartz, but didn't this used to work? Only way to get middle button is by setting emulate 3 buttons and command-left-click, which I don't remember doing before. I'm starting from a completely fresh install.

**Comment by nbriggs** [comment]
make sure you have the lispusers/WHEELSCROLL.LCOM loaded, and that `xev -event button` shows the expected button events (it should show up/down and left/right scroll as well as left/middle/right click)

**Comment by Anzus** [comment]
Ahh, rev -event reports button5 for the middle button. I think it should be button 3?

**Comment by Anzus** [comment]
I got that wrong 4 and 5 are when I scroll. Middle button doesn't seem to be doing anything.

**Comment by nbriggs** [comment]
Buttons:
1 - LEFT, 2 - MIDDLE, 3 - RIGHT,
4 - Keyset LEFT, 5 - Keyset LEFT-MIDDLE, 6 - Keyset RIGHT, 7 - Keyset RIGHT-MIDDLE

The mac trackpad generates:
  single-finger click: button 1; two-finger click: button 3 (if you have this configured in the system preferences)
  slide fingers up - button 5; slide fingers down - button 4
  slide fingers left - button 7; slide fingers right - button 6

I have "option-left-click" defined to be middle button, and emulate-3-button mouse on.

**Comment by nbriggs** [comment]
Check `man Xquartz` for the things it understands in `defaults write ...` for setting which modifier gives you button 2.

**Comment by Anzus** [comment]
Ha! Turns out it's hardware failure. Tried another trackball, it's generating button 2.

**Reply by nbriggs** [reply]
OK then!   I just connected a Kensington pocket mouse, and it works fine for 1/2/3 and up/down scroll (though it's a little hard to get a middle click without a scroll!)

**Reply by Anzus** [reply]
I have an "office trackball" that I leave in the office, and an "other" one that wanders with me. It was the "other" one that seems to have a busted middle click. Thanks for the help.

**Comment by rmkaplan** [comment]
It would be nice if we could figure out how to discriminate between an intended scroll and an accidental rotation with an intended button press.  There’s an issue on that, I think.

**Comment by masinter** [comment]
I'd like to have some strategy for how we will deal with different devices. I can't see any way to do much without a keyboard, but maybe:
- ab tablets with keyboard but no mouse (scroll bars?)
- trackpad (one "button" on click, but tends to lose position when you click
- one-button mouse 

this is a "modernization" to deal with modern hardware. 

Can we use function keys?

**Comment by masinter** [comment]
the discrimination between intended scroll and intended button press might depend on the hardware of the mouse.
We might add some kind of delay wheelscroll to ignore scrolls within N ms of a middle-button press, but there's already too much latency (Bluetooth mouse wake-up ) and I can even imagine some drivers doing this already.

i don't think it would be too onerous to add some kind of middle-mouse alternative ctrl-alt-space
